File: Silver/1932.cpp
// 제목 : 정수 삼각형
// 티어 : 실버 1
// 링크 : https://www.acmicpc.net/problem/1932
// 메모리(kb) : 2944
// 실행시간(ms) : 44

#define _CRT_SECURE_NO_WARNINGS 
#include<iostream>

using namespace std;

int main(int argc, char** argv)
{
    //삼각형의 크기
    int n;
    cin >> n;

    //삼각형을 저장할 배열을 할당하고 초기화
    int** arr = new int*[n + 1];
    for (int i = 0; i <= n; i++) {
        arr[i] = new int[n + 1];

    }

    for (int i = 0; i <= n; i++) {
        for (int j = 0; j <= n; j++) {
            arr[i][j] = 0;
        }
    }

    //배열에 값을 입력받으며, 그 위치에는 지금까지 선택한 수의 합 중 최댓값을 저장
    for (int i = 1; i <= n; i++)
        for (int j = 1; j <= i; j++) {
            int value;
            cin >> value;
            arr[i][j] = value + max(arr[i - 1][j - 1], arr[i - 1][j]);
        }

    //마지막 줄에서 최댓값을 찾은 후 출력
    int answer = 0;
    for (int j = 1; j <= n; j++) answer = answer > arr[n][j] ? answer : arr[n][j];

    cout << answer;
    return 0;
}

File: Silver/1932.java
// 제목 : 정수 삼각형
// 티어 : 실버 1
// 링크 : https://www.acmicpc.net/problem/1932
// 메모리(kb) : 25940
// 실행시간(ms) : 312

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

class Main {

	public static void main(String args[]) throws IOException {
		// 입출력 도구
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;

		//삼각형의 높이
		int n = Integer.parseInt(br.readLine());
		
		//삼각형 입력받기
		int[][] table = new int[n+1][n+1];
		for(int i=1;i<=n;i++) {
			st = new StringTokenizer(br.readLine());
			for(int j=1;j<=i;j++) {
				table[i][j] = Integer.parseInt(st.nextToken());
				//위의 두 값 중 큰 값을 더해줌
				table[i][j] += Math.max(table[i-1][j], table[i-1][j-1]);
			}
		}
		
		//마지막줄에서 최댓값 구하기
		int answer = 0;
		for(int i=1;i<=n;i++)
			answer = Math.max(answer, table[n][i]);
		
		//출력
		System.out.println(answer);
	}
}